[dynamo] recursively skip frames when Dynamo cache limit is hit

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135144

Fixes https://github.com/pytorch/pytorch/issues/128954 and [T197117723](https://www.internalfb.com/intern/tasks/?t=197117723).

In general, adds `SkipCodeRecursiveException` to Dynamo - when raised in Dynamo, convert_frame will return a `skip_code_recursive_flag` back to C Dynamo, signaling it to skip the current frame and all recursive calls.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

williamwen42 committed:
[dynamo] recursively skip frames when Dynamo cache limit is hit

[ghstack-poisoned]

williamwen42 commented:
@ezyang if the cache limit is hit, we will run into a similar problem to the one encountered in https://github.com/pytorch/pytorch/pull/132926: any downstream ops that should be force compiled will be skipped.

williamwen42 committed:
Update on "[dynamo] recursively skip frames when Dynamo cache limit is hit"


Fixes https://github.com/pytorch/pytorch/pull/135144 and [T197117723](https://www.internalfb.com/intern/tasks/?t=197117723).

In general, adds `SkipCodeRecursiveException` to Dynamo - when raised in Dynamo, the current frame and all nested frames will be unconditionally skipped.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -580,6 +585,13 @@ static PyObject* _custom_eval_frame(
   }
+  if (extra == SKIP_CODE_RECURSIVE) {
+    DEBUG_TRACE("skip recursive %s", get_frame_name(frame));
+    skip_code_recursive = true;
```
Why do we need a special flag for this?  Can't we just disable dynamo temporarily?

williamwen42 replied:
If we use `eval_frame_callback_set(Py_None);`, then any nested calls with a @compile decorator will re-enable tracing. Having a separate flag forces the eval frame callback to continue to be disabled in this case.

williamwen42 replied:
We will match the skip code recursive semantics with the disable decorator for now (will update when the disable decorator semantics change).

ezyang commented:
Yeah, it does kind of feel like calling into disable() (with the new semantics) is what you want here

anijain2305 commented:
Maybe I am missing something but isn't this what we want to do

```

if (skipped_due_to_recompilation) {

  // keep the callback None - disable for this frame and its children
  eval_frame_callback_set(Py_None);
  out = eval_frame_default(tstate, frame, throw_flat);

  // now renable
  eval_frame_callback_set(callback);
}
```

And this piece of code goes in this else branch - https://github.com/pytorch/pytorch/blob/main/torch/csrc/dynamo/eval_frame.c#L689-L697

And instead of raising exception and catching it, we might just use another code.

williamwen42 committed:
Update on "[dynamo] recursively skip frames when Dynamo cache limit is hit"


Fixes https://github.com/pytorch/pytorch/pull/135144 and [T197117723](https://www.internalfb.com/intern/tasks/?t=197117723).

In general, adds `SkipCodeRecursiveException` to Dynamo - when raised in Dynamo, the current frame and all nested frames will be unconditionally skipped.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] recursively skip frames when Dynamo cache limit is hit"


Fixes https://github.com/pytorch/pytorch/pull/135144 and [T197117723](https://www.internalfb.com/intern/tasks/?t=197117723).

In general, adds `SkipCodeRecursiveException` to Dynamo - when raised in Dynamo, convert_frame will return a `skip_code_recursive_flag` back to C Dynamo, signaling it to skip the current frame and all recursive calls.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -850,7 +853,13 @@ def format_guard_failures() -> str:
             )
-            unimplemented(f"{limit_type} reached")
+            if justknobs_check(
+                "pytorch/compiler:skip_code_recursive_on_cache_limit_hit"
```
Do you want to have a flag in config as well, just in case?

williamwen42 replied:
Added!

anijain2305 reviewed (approved):
Awesome! LGTM

williamwen42 committed:
Update on "[dynamo] recursively skip frames when Dynamo cache limit is hit"


Fixes https://github.com/pytorch/pytorch/pull/135144 and [T197117723](https://www.internalfb.com/intern/tasks/?t=197117723).

In general, adds `SkipCodeRecursiveException` to Dynamo - when raised in Dynamo, convert_frame will return a `skip_code_recursive_flag` back to C Dynamo, signaling it to skip the current frame and all recursive calls.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge
